### ISSUE TITLE
Firebase Cloud Messaging "notification" attribute conflicts with different platforms

### ISSUE THREAD

**Issue by Veeksi** [firebase/firebase-android-sdk#4204]
In cloud messaging server can send JSON payload like this to firebase that delivers notification with specified topic:
```json
{
    "to": "/topics/weather",
    "content_available": true,
    "mutable_content": true,
    "notification": {
        "title": "Test title",
        "body": "Test body",
        "image": "https://i.imgur.com/VakXt3Z.jpeg",
        "click_action": "OPEN_NEWS_ACTIVITY"
    },
    "data": {
        "title": "Test title",
        "body": "Test body",
        "image": "https://i.imgur.com/VakXt3Z.jpeg",
        "customField": "2022-03-24T00:00:00.000Z",
        "topic": "/topics/news-en"
    }
}
```
If user uses android, this json body triggers `onMessageReceived` callback only if app is on foreground or background. It doesn't trigger that callback if app is killed. In that case it triggers system tray that shows notification and all `notification` dictionary data gets lost when clicking notification. If we want to trigger `onMessageReceived` callback everytime, we could delete that `notification` dictionary so `onMessageReceived` callback gets called everytime. 

In iOS, `notification` dictionary is required to trigger any kind of notification. That's why if we want to send one notification to that topic, we cannot delete that `notification` dictionary.

So basically, we must duplicate data to satisfy both platforms and android side even requires more code because intent must be handled differently because onMessageReceived never gets called if the app is in wrong state. 

The question is: 
1. Is there way to trigger notification on iOS without providing `notification` attribute so `onMessageReceived` callback on android gets called everytime?
2. Is there way to trigger `onMessageReceived` callback on android even though we provide `notifications` dictionary that is needed on iOS?

**Comment by argzdev** [comment]
Hi @Veeksi, have you considered using the HTTP v1? It provides platform-specific blocks of keys.

e.g. 
```
{
  "message": {
    "topic": "news",
    "notification": {
      "title": "Breaking News",
      "body": "New news story available."
    },
    "data": {
      "story_id": "story_12345"
    },
    "android": {
      "notification": {
        "click_action": "TOP_STORY_ACTIVITY",
        "body": "Check out the Top Story"
      }
    },
    "apns": {
      "payload": {
        "aps": {
          "category" : "NEW_MESSAGE_CATEGORY"
        }
      }
    }
  }
}
```

I think this would fit your use case very well. You can check out our docs [here](https://firebase.google.com/docs/cloud-messaging/migrate-v1#example-simple-notification-message). Let me know if this works for you. Thanks!

**Comment by gsakakihara** [comment]
> It doesn't trigger that callback if app is killed. In that case it triggers system tray that shows notification and all notification dictionary data gets lost when clicking notification. 

https://firebase.google.com/docs/cloud-messaging/android/receive#backgrounded
If the user clicks the notification, the data payload should be included in the Intent extras of the Activity that was started.

If that doesn't work for you and you want to send a data message on Android that shows up as a notification on iOS, you may be able to specify the apns parameters for a notification, similar to argzdev's example. I'm not familiar with the APNs API, but something like this:

```
{
  "message": {
    "data": {
      ...
    },
    "apns": {
      "payload": {
        "aps": {
          "alert" : {
            "title" : "Test title",
            "body" : "Test body"
          },
          "category" : "NEW_MESSAGE_CATEGORY"
        }
      }
    }
  }
}
```

This also is similar to https://github.com/firebase/firebase-android-sdk/issues/1807 and maybe https://github.com/firebase/firebase-android-sdk/issues/1663.

**Comment by Veeksi** [comment]
> > It doesn't trigger that callback if app is killed. In that case it triggers system tray that shows notification and all notification dictionary data gets lost when clicking notification.
> 
> https://firebase.google.com/docs/cloud-messaging/android/receive#backgrounded If the user clicks the notification, the data payload should be included in the Intent extras of the Activity that was started.
> 
> If that doesn't work for you and you want to send a data message on Android that shows up as a notification on iOS, you may be able to specify the apns parameters for a notification, similar to argzdev's example. I'm not familiar with the APNs API, but something like this:
> 
> ```
> {
>   "message": {
>     "data": {
>       ...
>     },
>     "apns": {
>       "payload": {
>         "aps": {
>           "alert" : {
>             "title" : "Test title",
>             "body" : "Test body"
>           },
>           "category" : "NEW_MESSAGE_CATEGORY"
>         }
>       }
>     }
>   }
> }
> ```
> 
> This also is similar to #1807 and maybe #1663.

Yes, I have checked that APNs API but couldn't manage to trigger iOS notifications without that "notification" dictionary, sadly. 

And yes, currently I am handling that data payload by providing "click_action" if the notification is on system tray. But like I said above, it would be cool to always use that `onMessageReceived` callback whether there is notification dictionary or not.

**Comment by argzdev** [comment]
Thanks for these details, @Veeksi. Those does sound like reasonable use cases. 

Since this is similar to #1807, we can close this issue now to avoid duplication in tracking. Please do add an emoji thumbs up on the post (1807), so our engineers can prioritize adding this to the roadmap depending on the requests, and demand.
